### PR TITLE
R2s return source particle

### PR DESCRIPTION
**Pull request by zxkjack123**
Major change:
1. Add class `SourceParticle` in `pyne/src/source_sampling`.  
2. Change the return value of `Sampler::particle_birth` from `vector<double>` to `SourceParticle` object.

### CODE REVIEW

File: pyne/source_sampling.pxd [hunk 1]
```diff
@@ -31,7 +31,10 @@ cdef class AliasTable:
     cdef public np.ndarray _prob
     pass
 
+cdef class PySourceParticle:
+    cdef cpp_source_sampling.SourceParticle c_src
+    cdef public int c
```
Comment (gonuke) [inline pyne/source_sampling.pxd]: Why is `c` here but not others?

File: src/source_sampling.cpp [hunk 2]
```diff
@@ -92,7 +94,10 @@ std::vector<double> pyne::Sampler::particle_birth(std::vector<double> rands) {
   samp.push_back(pos[2]); 
   samp.push_back(sample_e(e_idx, rands[5]));
   samp.push_back(sample_w(pdf_idx));
-  return samp;
+  samp.push_back(-1.0);
```
Comment (gonuke) [inline src/source_sampling.cpp]: There is no need for the vector `samp` anymore... just initialize `SourceParticle` directly from `pos`, etc

File: src/source_sampling.cpp [hunk 3]
```diff
@@ -394,3 +399,39 @@ int pyne::AliasTable::sample_pdf(double rand1, double rand2) {
   int i = (int) n * rand1;
   return rand2 < prob[i] ? i : alias[i];
 }
+
+pyne::SourceParticle::SourceParticle() {
+    this->x = -1.0;
```
Comment (gonuke) [inline src/source_sampling.cpp]: `this` pointer is unnecessary.  Also - see initialization lists

File: src/source_sampling.cpp [hunk 4]
```diff
@@ -394,3 +399,39 @@ int pyne::AliasTable::sample_pdf(double rand1, double rand2) {
...
+    this->y = -1.0;
+    this->z = -1.0;
+    this->e = -1.0;
+    this->w = -1.0;
+    this->c = -1;
+}
+
+pyne::SourceParticle::SourceParticle(double x, double y, double z,
```
Comment (gonuke) [inline src/source_sampling.cpp]: Better to use function arguments that are different names from member variables, and then
A. Avoid using `this` pointer, and
B. Use initialization lists

File: src/source_sampling.cpp [hunk 5]
```diff
@@ -394,3 +399,39 @@ int pyne::AliasTable::sample_pdf(double rand1, double rand2) {
...
+    this->e = e;
+    this->w = w;
+    this->c = c;
+}
+
+pyne::SourceParticle::~SourceParticle() {};
+
+std::vector<double> pyne::SourceParticle::get_src_xyzew() {
+    std::vector<double> xyzew;
+    xyzew.push_back(this->x);
```
Comment (gonuke) [inline src/source_sampling.cpp]: Don’t need `this` pointer

File: src/source_sampling.h [hunk 6]
```diff
@@ -88,6 +90,35 @@ namespace pyne {
...
+    /// \param c The cell number of the source particle
+    SourceParticle(double x,
+                   double y,
+                   double z,
+                   double e,
+                   double w,
+                   int c);
+    ~SourceParticle();
+
+    double x; // x coordinate
```
Comment (gonuke) [inline src/source_sampling.h]: If you are providing access functions for these, perhaps they can/should be private member?

File: src/source_sampling.cpp [hunk 7]
```diff
@@ -394,3 +399,39 @@ int pyne::AliasTable::sample_pdf(double rand1, double rand2) {
...
+    xyzew.push_back(this->y);
+    xyzew.push_back(this->z);
+    xyzew.push_back(this->e);
+    xyzew.push_back(this->w);
+    return xyzew;
+}
+
+int pyne::SourceParticle::get_src_c() {
```
Comment (gonuke) [inline src/source_sampling.cpp]: Can probably inline this in header.  Don’t need `this` pointer

**Review by gonuke — CHANGES_REQUESTED** [review]
This is mostly ok.  Just a few cleanup suggestions.

**Comment by zxkjack123** [comment]
@gonuke Clean up the code acoording to your suggestions. 
Major changes:
1. Remove unsued definition of `int c` in `source_sampling.pxd`.
2. Remove `samp` of `particle_birth`, use data to initialize `SourceParticle`. 
3. Remove `this` point in constructors of `SourceParticle`.

**Comment by zxkjack123** [comment]
Update:
1. Make x, y, z, e, w, c of `SourceParticle` private. 
2. Rename `get_src_xxx` into `get_xxx`, make shorter function names.

File: pyne/cpp_source_sampling.pxd [hunk 8]
```diff
@@ -30,6 +30,24 @@ cdef extern from "source_sampling.h" namespace "pyne":
...
+
+        # methods
+        cpp_vector[double] get_xyzew() except +
+        double get_x() except +
+        double get_y() except +
+        double get_z() except +
+        double get_e() except +
+        double get_w() except +
+        int get_c() except +
+        pass
```
Comment (scopatz) [inline pyne/cpp_source_sampling.pxd]: You don't need the `pass` here

File: pyne/cpp_source_sampling.pxd [hunk 9]
```diff
@@ -47,11 +65,11 @@ cdef extern from "source_sampling.h" namespace "pyne":
...
+        SourceParticle particle_birth() except +
+        SourceParticle particle_birth(cpp_vector[double]) except +
         pass
 
 
 
 
-{'cpppxd_footer': '', 'pyx_header': '', 'pxd_header': '', 'pxd_footer': '', 'cpppxd_header': '', 'pyx_footer': ''}
\ No newline at end of file
+{'cpppxd_footer': '', 'pyx_header': '', 'pxd_header': '', 'pxd_footer': '', 'cpppxd_header': '', 'pyx_footer': ''}
```
Comment (scopatz) [inline pyne/cpp_source_sampling.pxd]: This line could probably be removed.  Not sure what it is doing here :)

File: pyne/source_sampling.pxd [hunk 10]
```diff
@@ -31,7 +31,9 @@ cdef class AliasTable:
...
 
+cdef class PySourceParticle:
+    cdef cpp_source_sampling.SourceParticle c_src
+    pass
 
 
-
-{'cpppxd_footer': '', 'pyx_header': '', 'pxd_header': '', 'pxd_footer': '', 'cpppxd_header': '', 'pyx_footer': ''}
\ No newline at end of file
+{'cpppxd_footer': '', 'pyx_header': '', 'pxd_header': '', 'pxd_footer': '', 'cpppxd_header': '', 'pyx_footer': ''}
```
Comment (scopatz) [inline pyne/source_sampling.pxd]: This could also be removed

File: pyne/source_sampling.pxd [hunk 11]
```diff
@@ -31,7 +31,9 @@ cdef class AliasTable:
     cdef public np.ndarray _prob
     pass
 
+cdef class PySourceParticle:
+    cdef cpp_source_sampling.SourceParticle c_src
+    pass
```
Comment (scopatz) [inline pyne/source_sampling.pxd]: This pass is not needed.  You may want to remove the others in the file.

File: pyne/source_sampling.pyx [hunk 12]
```diff
@@ -478,14 +477,158 @@ cdef class Sampler:
...
+        """
+        cdef cpp_vector[double] rtnval
+        cdef np.npy_intp rtnval_proxy_shape[1]
+        rtnval = self.c_src.get_xyzew()
         rtnval_proxy_shape[0] = <np.npy_intp> rtnval.size()
         rtnval_proxy = np.PyArray_SimpleNewFromData(1, rtnval_proxy_shape, np.NPY_FLOAT64, &rtnval[0])
         rtnval_proxy = np.PyArray_Copy(rtnval_proxy)
         return rtnval_proxy
+
+    def get_x(self):
```
Comment (scopatz) [inline pyne/source_sampling.pyx]: I think that you should use properties here rather than getters for all of these

Comment (scopatz) [reply pyne/source_sampling.pyx]: Basically you just need to do, 

```python
@property
def x(self):
    ...
```

File: pyne/source_sampling.pyx [hunk 13]
```diff
@@ -478,14 +477,158 @@ cdef class Sampler:
...
+    y (double) : The y coordinate of the source particle
+    z (double) : The z coordinate of the source particle
+    e (double) : The energy of the source particle
+    w (double) : The weight of the source particle
+    c (int) : The cell number of the source particle
+    
+    Methods
+    -------
+    SourceParticle
+    ~SourceParticle
```
Comment (scopatz) [inline pyne/source_sampling.pyx]: You don't need to list the methods like this, especially not the constructor/destructor.

File: pyne/source_sampling.pyx [hunk 14]
```diff
@@ -478,14 +477,158 @@ cdef class Sampler:
             rands_proxy = cpp_vector[double](<size_t> rands_size)
             for irands in range(rands_size):
                 rands_proxy[irands] = <double> rands[irands]
-        rtnval = (<cpp_source_sampling.Sampler *> self._inst).particle_birth(rands_proxy)
+        cdef cpp_source_sampling.SourceParticle c_src = (<cpp_source_sampling.Sampler *> self._inst).particle_birth(rands_proxy)
+        return PySourceParticle(c_src.get_x(), c_src.get_y(), c_src.get_z(), c_src.get_e(), c_src.get_w(), c_src.get_c())
+
+
+
+cdef class PySourceParticle:
```
Comment (scopatz) [inline pyne/source_sampling.pyx]: You should probably remove the `Py` part of the name

Comment (scopatz) [reply pyne/source_sampling.pyx]: There is no reason for them to have different names in Python and C++

**Comment by scopatz** [comment]
Thanks for putting this in @zxkjack123!

**Comment by zxkjack123** [comment]
Update: some clean up and improvement according to the suggestions of @scopatz .
Major changes:
1. Remove the `Py` from `PySourceParticle`. 
2. Get `c` using property rather than getter.

**Comment by gonuke** [comment]
I thought you needed the `PySourceParticle` to avoid some kind of build problem?

File: pyne/source_sampling.pyx [hunk 15]
```diff
@@ -478,20 +477,63 @@ cdef class Sampler:
             rands_proxy = cpp_vector[double](<size_t> rands_size)
             for irands in range(rands_size):
                 rands_proxy[irands] = <double> rands[irands]
-        rtnval = (<cpp_source_sampling.Sampler *> self._inst).particle_birth(rands_proxy)
-        rtnval_proxy_shape[0] = <np.npy_intp> rtnval.size()
-        rtnval_proxy = np.PyArray_SimpleNewFromData(1, rtnval_proxy_shape, np.NPY_FLOAT64, &rtnval[0])
-        rtnval_proxy = np.PyArray_Copy(rtnval_proxy)
-        return rtnval_proxy
+        cdef cpp_source_sampling.SourceParticle c_src = (<cpp_source_sampling.Sampler *> self._inst).particle_birth(rands_proxy)
+        return SourceParticle(c_src.get_x(), c_src.get_y(), c_src.get_z(), c_src.get_e(), c_src.get_w(), c_src.get_c())
```
Comment (scopatz) [inline pyne/source_sampling.pyx]: Lines too long

File: pyne/source_sampling.pyx [hunk 16]
```diff
@@ -478,20 +477,63 @@ cdef class Sampler:
...
+        """
+        cdef cpp_vector[double] rtnval
+        cdef np.npy_intp rtnval_proxy_shape[1]
+        rtnval = self.c_src.get_xyzew()
+        rtnval_proxy_shape[0] = <np.npy_intp> rtnval.size()
+        rtnval_proxy = np.PyArray_SimpleNewFromData(1, rtnval_proxy_shape, np.NPY_FLOAT64, &rtnval[0])
+        rtnval_proxy = np.PyArray_Copy(rtnval_proxy)
+        return rtnval_proxy
 
+    property c:
```
Comment (scopatz) [inline pyne/source_sampling.pyx]: Please use new style properties: http://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#properties

**Comment by zxkjack123** [comment]
@gonuke Acutally, I haven't figured out the exact reason that caused the build problem with `PySourceParticle` before. I changed several things, including the name `PySourceParticle`. When it finally worked well, I thought it was the reason of the name `PySourceParticle`. 
However, it worked too when I tried to remove the `Py`. Now I see it's not because of the name, but maybe (still not sure) the reason of `cpp_source_sampling`.

File: pyne/source_sampling.pyx [hunk 17]
```diff
@@ -478,20 +477,63 @@ cdef class Sampler:
...
+    y (double) : The y coordinate of the source particle
+    z (double) : The z coordinate of the source particle
+    e (double) : The energy of the source particle
+    w (double) : The weight of the source particle
+    c (int) : The cell number of the source particle
     
+    Methods
+    -------
+    SourceParticle
+    ~SourceParticle
```
Comment (scopatz) [inline pyne/source_sampling.pyx]: Please remove the methods section here

File: pyne/source_sampling.pyx [hunk 18]
```diff
@@ -478,20 +478,65 @@ cdef class Sampler:
...
+        rtnval = self.c_src.get_xyzew()
+        rtnval_proxy_shape[0] = <np.npy_intp> rtnval.size()
+        rtnval_proxy = np.PyArray_SimpleNewFromData(1, rtnval_proxy_shape, np.NPY_FLOAT64, &rtnval[0])
+        rtnval_proxy = np.PyArray_Copy(rtnval_proxy)
+        return rtnval_proxy
 
-{'cpppxd_footer': '', 'pyx_header': '', 'pxd_header': '', 'pxd_footer': '', 'cpppxd_header': '', 'pyx_footer': ''}
+    @property
+    def c(self):
+        return self.c_src.get_c()
```
Comment (scopatz) [inline pyne/source_sampling.pyx]: Don't you also want properties for x, y, z, w, e?

Comment (zxkjack123) [reply pyne/source_sampling.pyx]: No, I don't need x, y, z, w, e separately. They will be accessed in a vector that contains all of them.

**Comment by zxkjack123** [comment]
Update: Removed some comments according to the suggestion of @scopatz .

File: pyne/source_sampling.pyx (hunk 18 above)
Comment (zxkjack123) [reply pyne/source_sampling.pyx]: I find that add properties for x, y, z, w and e will make test functions more readable. I add these properties and modified test functions.

**Comment by zxkjack123** [comment]
Update: Add properties x, y, z, e, w. Change variable name in test functions of test_source_sampling.py. Use `s.x` to replace `s[0]`. Make the test code easier to understand.

File: pyne/source_sampling.pyx [hunk 19]
```diff
@@ -478,20 +478,77 @@ cdef class Sampler:
...
+        return SourceParticle(c_src.get_x(), c_src.get_y(), c_src.get_z(), \
+                c_src.get_e(), c_src.get_w(), c_src.get_c())
+
+
+cdef class SourceParticle:
+    """Constructor for class SourceParticle
+    
+    Attributes
+    ----------
+    x (double) : The x coordinate of the source particle
```
Comment (scopatz) [inline pyne/source_sampling.pyx]: Please follow numpydoc format, these would be:

```
x : double
    The x coordinate of the source particle
```

**Comment by scopatz** [comment]
Thanks @zxkjack123! This is looking a lot better! Just the one comment about the docstring.

**Comment by zxkjack123** [comment]
Update: Cleaned up some documentations according to the suggestions of @scopatz .

**Review by scopatz — APPROVED** [review]

**Comment by scopatz** [comment]
LGTM other than the build failure :)

**Review by gonuke — APPROVED** [review]

**Comment by gonuke** [comment]
Thanks @zxkjack123

**Pull request merged**